### PR TITLE
bump crengine: tables rendering improvements and others 

### LINKED ISSUE #2846: CSS: table: vertical-align; td: width unsupported
* KOReader version: v2015.11-1024-g3b813c7
* Device: Emulator

#### Issue
The following basic formatting for tables seems to be unsupported:
`vertical-align` seems to be ignored for `table` and `td` elements
`width` seems to be ignored for `td` elements

### LINKED ISSUE #4196: Problem with a symbol using Crengine
Hey, I have a book which has a symbol that crengine doesn't display. It is fine if I use a ebook reader at PC.

Reproduce => Go to ( "capítulo I Estado fiscal, liberdade e tributação" and then "3 As funções fiscais")

COMPUTER

computer

KOREADER
reader_2018-sep-04_085905

Is there a way to fix it?

### PR DESCRIPTION
**Pull request by poire-z**
bump crengine, which includes https://github.com/koreader/crengine/pull/243 and https://github.com/koreader/crengine/pull/245 : 
- Adds support for symbol fonts (local or embedded) (See https://github.com/koreader/koreader/issues/4196#issuecomment-435587002 Closes #4196.)
- Fix some issues when rendering text in constrained width
- Page splitting: fix possible missing blocks
- CSS: adds support for 'auto', ignore % for borders
- Fix right border drawing position
- Fix: adds missing properties in copystyle()
- Adds comments, erm_killed rendering method
- Adds getRenderedWidths(): get node min/max node content width
- Tables rendering: fixes and improvements (Closes #2846, vertical-align now supported)
- getRenderedWidths: enable min_width to be a single CJK char
- Fix wrong text wrap avoid in some case
- epub.css: add style for 'blockquote'
- Fix rendering issue when line ends with an image

Adds a few style tweaks related to tables (corrections or alternate english wording welcome :)

Enforce table width: 100% in Wikipedia EPUBs to keep previous look, which feels better with the various kinds of tables in Wikipedia pages.

Also includes: Bump FBInk to 1.9.2 (https://github.com/koreader/koreader-base/pull/770) Bump libpng to 1.6.36 https://github.com/koreader/koreader-base/pull/771 and some android build tweaks https://github.com/koreader/koreader-base/pull/750

Also includes a fix for diagonal swipe to refresh full screen not working in TOC and History (see https://github.com/koreader/koreader/pull/4240#issuecomment-443511635).

### CODE REVIEW

**Comment by poire-z** [comment]
Oh, joy, unit test failures... :(

**Comment by poire-z** [comment]
Oh my, juliet.epub is full of `<blockquote>` ... so it goes from 189 pages to 262 pages...

Old rendering:
<kbd>image</kbd>

New rendering:
<kbd>image</kbd>

Do we agree the new is better than the old and worth the 40% raise in pages count?

(I'd rather drop the blockquote style than have to adapt all these unit tests :) ... but well, if I have to...)

An other small thing is that the cover page does and did overflow the page height. But in the page splitting code, a `SplitLineIfOverflowPage(line)` in the [first page handling code](https://github.com/koreader/crengine/blob/0d5dc2d9f7a5d636735d9b43e15479025f2f3f1f/crengine/src/lvpagesplitter.cpp#L316-L323) was missing, that I added.
So, on my 578px page height emulator, the first page is 583px, and the 5px were previously just dropped. Now, they are pushed to next page, which is now a blank page (the cover image is rightly 578px, but because of the "strut", 5px below the baseline (from the DIV font size) is added below the image (and it is the right behaviour that if there are these 5px, they have to not be dropped and pushed to next page):
<kbd>image</kbd>

Now, there is another small thing: I didn't read it in the spec, but Firefox seems to _not_ enforce this strut when there is no text (or no non-space-only text) alongside an image:
```html
<div style="background-color: #ff0000">
<img src="Lac_sevan_armenie.png">
</div>
```
<kbd>image</kbd>
Add a single letter (`e`, or `"` here) in the DIV, and the strut is enforced:
<kbd>image</kbd>

<kbd>image</kbd>
with:
```html
<div style="background-color: #ff0000; font-size: 400%">
<img src="Lac_sevan_armenie.png"> "
</div>
```

crengine currently enforces it even when there is no text:
<kbd>image</kbd>
(i see no quick and easy solution to have it does as Firefox - so we'd have to do with a blank page following an inline image with height 100%)

**Comment by poire-z** [comment]
Additional note: Firefox does as crengine does when I add at the top of the HTML
```html
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.1//EN"
  "http://www.w3.org/TR/xhtml11/DTD/xhtml11.dtd">
```
image

When the above "html header/doctype" is missing, all the [strut rules stated  by the specs](https://www.w3.org/TR/CSS2/visudet.html#line-height) (that I added in the [vertical-align PR ](https://github.com/koreader/crengine/pull/240) ) are not enforced by Firefox...

So, essential question: is it ok/fine/best that crengine _now_ does what Firefox seems to do only when .`..DTD/xhtml11` ??

It's really by chance I had this HTML "header" in my test file when I did the vertical-align PR...

**Comment by poire-z** [comment]
Checked firefox with various doctypes mentionned in https://www.w3schools.com/tags/tag_doctype.asp.

strut is enforced with
```
<!DOCTYPE html>
<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01//EN" "http://www.w3.org/TR/html4/strict.dtd">
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.1//EN" "http://www.w3.org/TR/xhtml11/DTD/xhtml11.dtd">
```

strut is NOT enforced with
```
<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Frameset//EN" "http://www.w3.org/TR/html4/frameset.dtd">
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
```

(We won't implement different behaviousr based on the doctype in crengine :) we should just have a single most adequate one.)

**Comment by NiLuJe** [comment]
I kinda prefer the new bq rendering, yeah ;).

**Comment by Frenzie** [comment]
I agree that the new rendering is more like how one would expect a blockquote to render. Otherwise you'd barely know the difference with a paragraph.

calibre for comparison:

screenshot_2018-12-06_20-28-48

Regarding the strut, you just found out that in between quirks mode and standards mode, there's almost standards mode. ;-)

https://developer.mozilla.org/en-US/docs/Mozilla/Gecko_Almost_Standards_Mode

Particularly [here](https://developer.mozilla.org/en-US/docs/Mozilla/Gecko_Almost_Standards_Mode#Triggering_.22Almost_Standards.22).

**Comment by Frenzie** [comment]
> (We won't implement different behaviousr based on the doctype in crengine :) we should just have a single most adequate one.)

The current behavior is excellent.

**Comment by poire-z** [comment]
Thanks, that explains it all. And thanks for the positives.
Any comment about the style tweaks wording, before I merge this? (I'll look in the mathml/image degradation soon, hoping it does not degrade anything else more classic).

**Comment by Frenzie** [comment]
Oh, my bad. I thought it was just the crengine bump. I'll have a look in an hour.

File: frontend/ui/data/css_tweaks.lua [hunk 1]
```diff
@@ -191,40 +191,72 @@ h1, h2, h3, h4, h5, h6 { hyphens: none !important; }
     {
         title = _("Miscellaneous"),
         {
-            id = "table_row_odd_even";
-            title = _("Alternate background color of table rows"),
-            css = [[
+            title = _("Tables"),
+            {
+                id = "table_full_width";
+                title = _("Make tables use the full width"),
```
Comment (Frenzie) [inline frontend/ui/data/css_tweaks.lua]:
```suggestion
                title = _("Full-width tables"),
```

File: frontend/ui/data/css_tweaks.lua [hunk 2]
```diff
@@ -191,40 +191,72 @@ h1, h2, h3, h4, h5, h6 { hyphens: none !important; }
...
+                description = _("Have table use the full width of the page. (Tables with small content now uses only the width needed to display that content. This restores the previous behaviour.)"),
```
Comment (Frenzie) [inline frontend/ui/data/css_tweaks.lua]:
```suggestion
                description = _("Make table expand to the full width of the page. (Tables with small content now use only the width needed to display that content. This restores the previous behavior.)"),
```

(Regarding behavior, I think we were going for a more AmE speling on the whole.)

File: frontend/ui/data/css_tweaks.lua [hunk 3]
```diff
@@ -191,40 +191,72 @@ h1, h2, h3, h4, h5, h6 { hyphens: none !important; }
...
+                css = [[table { width: 100% !important; }]],
+            },
+            {
+                id = "table_td_width_auto";
+                title = _("Ignore table and cells specified widths"),
```
Comment (Frenzie) [inline frontend/ui/data/css_tweaks.lua]:
```suggestion
                title = _("Ignore specified table and cell widths"),
```

author … widths?
document … widths?
publisher … widths?

File: frontend/ui/data/css_tweaks.lua [hunk 4]
```diff
@@ -191,40 +191,72 @@ h1, h2, h3, h4, h5, h6 { hyphens: none !important; }
...
+                description = _("Ignore table and cells widths specified by publisher, and let the engine decide for the most adequate widths."),
```
Comment (Frenzie) [inline frontend/ui/data/css_tweaks.lua]:
```suggestion
                description = _("Ignore table and cells widths specified by the publisher, and let the engine decide the most appropriate widths."),
```

File: frontend/ui/data/css_tweaks.lua [hunk 5]
```diff
@@ -191,40 +191,72 @@ h1, h2, h3, h4, h5, h6 { hyphens: none !important; }
...
+                css = [[table, td, th { width: auto !important; }]],
+            },
+            {
+                id = "table_margin_left_right_auto";
+                title = _("Center small tables"),
+                description = _("Horizontally center tables that do not use the full page width"),
```
Comment (Frenzie) [inline frontend/ui/data/css_tweaks.lua]:
```suggestion
                description = _("Horizontally center tables that do not use the full page width."),
```

File: frontend/ui/data/css_tweaks.lua [hunk 6]
```diff
@@ -191,40 +191,72 @@ h1, h2, h3, h4, h5, h6 { hyphens: none !important; }
...
-            id = "image_full_width";
-            title = _("Make images full width"),
-            description = _("Useful for books containing only images, when they are smaller than your screen. May stretch images in some cases."),
-            -- This helped me once with a book. Will mess with aspect ratio
-            -- when images have a style="width: NNpx; heigh: NNpx"
-            css = [[
+            title = _("Images"),
+            {
+                id = "image_full_width";
+                title = _("Make images full width"),
```
Comment (Frenzie) [inline frontend/ui/data/css_tweaks.lua]:
```suggestion
                title = _("Full-width images"),
```

File: frontend/ui/data/css_tweaks.lua [hunk 7]
```diff
@@ -191,40 +191,72 @@ h1, h2, h3, h4, h5, h6 { hyphens: none !important; }
...
+                description = _("Useful for books containing only images, when they are smaller than your screen. May stretch images in some cases."),
+                -- This helped me once with a book. Will mess with aspect ratio
+                -- when images have a style="width: NNpx; heigh: NNpx"
```
Comment (Frenzie) [inline frontend/ui/data/css_tweaks.lua]: In your average browser you could just set height:auto !important.

**Pull request merged**